### PR TITLE
net/dns:fix unaligned access in dns query

### PR DESCRIPTION
**Pull request by Virus-V**
## Summary
fix unaligned access in dns query

### CODE REVIEW

File: libs/libc/netdb/lib_dnsquery.c [hunk 1]
```diff
@@ -439,8 +440,11 @@ static int dns_recv_response(int sd, FAR union dns_addr_u *addr, int naddr,
   ninfo("Question: type=%04x, class=%04x\n",
         ntohs(que->type), ntohs(que->class));
 
-  if (que->type  != qinfo->rectype ||
-      que->class != HTONS(DNS_CLASS_IN))
+  /* N.B. Unaligned access may occur here */
+
+  temp = HTONS(DNS_CLASS_IN);
+  if (memcmp(&que->type, &qinfo->rectype, sizeof(uint16_t)) != 0 ||
```
Comment (xiaoxiang781216) [inline libs/libc/netdb/lib_dnsquery.c:446]: But both dns_header_s and dns_question_s are naturally aligned, the unalignment most likely come from buffer. The better fix is ensure buffer variable start at the four bytes boundary.

Comment (Virus-V) [reply libs/libc/netdb/lib_dnsquery.c:446]: No, DNS request/response struct are packed, changing buffer type does not work.

Comment (xiaoxiang781216) [reply libs/libc/netdb/lib_dnsquery.c:446]: dns_header_s and dns_question_s don't have __attribute__((packed)), why you consider they are packed?

Comment (Virus-V) [reply libs/libc/netdb/lib_dnsquery.c:446]: dns_question_s is part of the dns response message. There is a variable length area in front of this structure. I am talking about the DNS response message, not the dns_question_s structure.

Comment (xiaoxiang781216) [reply libs/libc/netdb/lib_dnsquery.c:446]: Ok, but line 430-440 need be removed too.

Comment (Virus-V) [reply libs/libc/netdb/lib_dnsquery.c:446]: You can try to `ping www.baidu.com` and `ping www.baidui.com`, the second domain name will have one more byte, to compare the address of `que`.

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM, thanks.

**Pull request merged**